### PR TITLE
Correct @odata.context for combination of $select/$apply/$compute/$expand

### PR DESCRIPTION
**Pull request by kosinsky**
### Description

Fixes to @odata.context when $select/$expand/$apply/$compute work together:

* Support for $apply inside $expand
 For `$expand=Sales($apply=aggregate(Amount with sum as Total))` @odata.context will be `$metadata#Products(Sales(Total))` (https://issues.oasis-open.org/browse/ODATA-1307)
* $select after $apply and $compute keeps only fields in $select

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### CODE REVIEW

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ODataContextUriBuilderTests.cs [hunk 1]
```diff
@@ -165,7 +176,68 @@ public void FeedContextUriWithApply()
...
+            }
+        }
+
+        [Fact]
+        public void FeedContextUriWithApplyAndSelect()
+        {
+            string applyClause = "groupby((Name), aggregate(Id with sum as TotalId))";
+            foreach (ODataVersion version in Versions)
+            {
+                this.CreateFeedContextUri("Name", null, applyClause, null, version).OriginalString.Should().Be(MetadataDocumentUriString + "#Cities(Name)");
```
Comment (mikepizzo) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ODataContextUriBuilderTests.cs]: >#Cities(Name) [](start = 140, length = 13)

Would this not be #Cities(Name,TotalId)?

File: src/Microsoft.OData.Core/UriParser/SemanticAst/SelectExpandClauseExtensions.cs [hunk 2]
```diff
@@ -94,9 +95,10 @@ internal static List<string> GetCurrentLevelSelectList(this SelectExpandClause s
         /// <typeparam name="T">Type of the sub processing result for expand items.</typeparam>
         /// <param name="selectExpandClause">The select expand clause for evaluation.</param>
         /// <param name="processSubResult">The method to deal with sub expand result.</param>
-        /// <param name="combineSelectAndExpand">The method to combine select and expand result lists.</param>
+        /// <param name="combineSelectAndExpand">The method to deal with apply result.</param>
+        /// <param name="processApply">The method to combine select and expand result lists.</param>
```
Comment (mikepizzo) [inline src/Microsoft.OData.Core/UriParser/SemanticAst/SelectExpandClauseExtensions.cs]: are these two descriptions switched?

File: src/Microsoft.OData.Core/UriParser/SemanticAst/SelectExpandClauseExtensions.cs [hunk 3]
```diff
@@ -107,12 +109,16 @@ internal static void Traverse<T>(this SelectExpandClause selectExpandClause, Fun
                 T subResult = default(T);
                 if (expandSelectItem.SelectAndExpand.SelectedItems.Any())
                 {
-                    Traverse(expandSelectItem.SelectAndExpand, processSubResult, combineSelectAndExpand,
-                        out subResult);
+                    Traverse(expandSelectItem.SelectAndExpand, processSubResult, combineSelectAndExpand, processApply, out subResult);
                 }
 
-                var expandItem = processSubResult(currentExpandClause, subResult);
+                // TODO: That's PoC we need to handle bunch of cases.
```
Comment (mikepizzo) [inline src/Microsoft.OData.Core/UriParser/SemanticAst/SelectExpandClauseExtensions.cs]: Please open an issue to track the todo, and add the issue # to the comment.

File: src/Microsoft.OData.Core/UriParser/SemanticAst/SelectExpandClauseExtensions.cs (hunk 2 above)
Comment (kosinsky) [reply src/Microsoft.OData.Core/UriParser/SemanticAst/SelectExpandClauseExtensions.cs]: Yep.Fixed

File: src/Microsoft.OData.Core/UriParser/SemanticAst/SelectExpandClauseExtensions.cs (hunk 3 above)
Comment (kosinsky) [reply src/Microsoft.OData.Core/UriParser/SemanticAst/SelectExpandClauseExtensions.cs]: It's handled. Removed the comment

Comment (mikepizzo) [reply src/Microsoft.OData.Core/UriParser/SemanticAst/SelectExpandClauseExtensions.cs]: Excellent; thanks!

---
In reply to: [321509085](https://github.com/OData/odata.net/pull/1526#discussion_r321509085) [](ancestors = 321509085)

**Pull request merged**